BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c

```diff
@@ -54,49 +56,71 @@ int VerifyRsaSignature (CONST PUB_KEY_HDR *PubKeyHdr, CONST SIGNATURE_HDR *Signa
     return err;
   }
 
-  rsa_key_s = (IppsRSAPublicKeyState*)bn_buf;
-  sz_adj = ((sz_rsa + sizeof(Ipp32u)) & ~(sizeof(Ipp32u)-1));
-  scratch_buf = bn_buf + sz_adj;
-  sz_adj = sz_adj + sz_adj;
-  bn_rsa_n = (IppsBigNumState *)(bn_buf + sz_adj);
-  sz_adj = sz_adj + ((sz_n + sizeof(Ipp32u)) & ~(sizeof(Ipp32u)-1));
-  bn_rsa_e = (IppsBigNumState *)(bn_buf + sz_adj);
-  sz_adj = sz_adj + ((sz_e + sizeof(Ipp32u)) & ~(sizeof(Ipp32u)-1));
+  // Allign sz
+  sz_rsa = IPP_ALIGNED_SIZE (sz_rsa, sizeof(Ipp32u));
+  sz_n   = IPP_ALIGNED_SIZE (sz_n, sizeof(Ipp32u));
+  sz_e   = IPP_ALIGNED_SIZE (sz_e, sizeof(Ipp32u));
 
-  if (sz_adj > sizeof(bn_buf)) {
+  // Allocate BN Buf
+  bn_buf = AllocateTemporaryMemory (sz_rsa + sz_n + sz_e);
+  if (bn_buf ==  NULL) {
     return ippStsNoMemErr;
   }
 
+  bn_buf_ptr   = bn_buf;
+  rsa_key_s    = (IppsRSAPublicKeyState*) bn_buf_ptr;
+  bn_buf_ptr   = bn_buf_ptr + sz_rsa;
+  bn_rsa_n     = (IppsBigNumState *) bn_buf_ptr;
+  bn_buf_ptr   = bn_buf_ptr + sz_n;
+  bn_rsa_e     = (IppsBigNumState *) bn_buf_ptr;
+
   err = ippsBigNumInit(mod_len / sizeof(Ipp32u), bn_rsa_n);
   if (err != ippStsNoErr) {
+    FreeTemporaryMemory (bn_buf);

```

Comment:
can we have a better flow to handle error ?  too many duplicated free memory here.